tests/src/test/java/org/eclipse/hono/tests/client/HonoClientImplIT.java

```diff
@@ -0,0 +1,81 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2019 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.tests.client;
+
+import java.net.HttpURLConnection;
+
+import org.eclipse.hono.client.ServiceInvocationException;
+import org.eclipse.hono.config.ClientConfigProperties;
+import org.eclipse.hono.tests.IntegrationTestHonoClient;
+import org.eclipse.hono.tests.IntegrationTestSupport;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import io.vertx.core.Vertx;
+import io.vertx.ext.unit.Async;
+import io.vertx.ext.unit.TestContext;
+import io.vertx.ext.unit.junit.VertxUnitRunner;
+import io.vertx.proton.ProtonClientOptions;
+
+/**
+ * Test cases verifying the behavior of {@code HonoClient}.
+ */
+@RunWith(VertxUnitRunner.class)
+public class HonoClientImplIT {
+
+    private static Vertx vertx;
+
+    public IntegrationTestHonoClient honoClient;
+
+    /**
+     * Sets up vert.x.
+     */
+    @BeforeClass
+    public static void init() {
+        vertx = Vertx.vertx();
+    }
+
+    /**
+     * Verifies that a connection attempt where no credentials are given fails
+     * immediately with a ClientErrorException with status code 401.
+     * 
+     * @param ctx The vert.x test context.
+     */
+    @Test
+    public void testConnectFailsWithClientErrorForNoSASLMechanismException(final TestContext ctx) {
```

Comment:
IMHO we should move this test to `org.eclipse.hono.tests.amqp.AmqpConnectionIT` because
* we already perform similar tests there as well and
* the AMQP adapter is a component under our own control which we can make sure behaves in the way we expect it to

Reply:
At first glance, I also thought so - seeing the already existing `testConnectFailsForWrongCredentials` method there.
But, this method is using the `ProtonClient` directly instead of the `HonoClient`. Of course this could be changed, but that would mean either having this one case (plus the new test method) where the `AmqpAdapterTestBase.connectToAdapter` method is not used in favour of a HonoClient-using method. Or it would mean adapting all invocations of `AmqpAdapterTestBase.connectToAdapter` - these are quite many.

So, I think this test doesn't fit that well there after all, since the focus in the `amqp` package is on the AMQP adapter.

Having a separate `HonoClientImplIT` in a separate package makes it clearer that this test is about `HonoClient` functionality. And it also offers the possibility of adding various other test methods for testing the `HonoClientImpl` there later on.
If the test class is already there, the hurdle of actually adding an integration test later on is lower.

Reply:
I still don't like it. In particular, we are not verifying any specific behavior of Hono here, so I do not see why we should have an *integration test* for `HonoClientImpl`. FMPOV all of the client's expected behavior can easily be tested using unit tests (which you already do).
Anyway, I will merge it for the time being. Let's see how it evolves ...

Reply:
> FMPOV all of the client's expected behavior can easily be tested using unit tests

The thing here is that the client's behavior is also defined by the underlying vert.x-proton.
The plain unit test relies on mocks to mimic the behavior of vert.x proton and qpid. If that behavior changes with an update (e.g. throwing an `AuthenticationException` in `handleSaslInit`) the unit test possibly is still green. The integration test won't be in that case.